src/amber/cli/helpers/process_runner.cr

```diff
@@ -1,114 +1,217 @@
+      at_exit do
+        kill_processes
+      end
+
+      Signal::INT.trap do
+        Signal::INT.reset
+        exit
+      end
     end
 
     def run
-      loop do
-        scan_files
-        sleep 1
+      if watch_object = CLI.config.watch
+        run_watcher(watch_object)
+      else
+        error "Can't find watch settings, please check your .amber.yml file"
+        exit 1
       end
+    rescue ex : KeyError
+      error "Error in watch configuration. #{ex.message}"
+      exit 1
     end
 
-    # Compiles and starts the application
-    def start_app
-      build_result = build_app_process
-      if build_result.is_a? Process::Status
-        if build_result.success?
-          stop_all_processes
-          create_all_processes
-          @app_running = true
-        elsif !@app_running
-          log "Compile time errors detected. Shutting down..."
-          exit 1
+    private def run_watcher(watch_object)
+      watch_object.each do |key, value|
+        next if key == "client"
+        files = value["files"]
+        commands = value["commands"]
+        if key != "server"
+          @notify_counter += 1
         end
+        spawn watcher(key, files, commands)
       end
+      @notify_counter_channel.send @notify_counter
+      @notify_counter = @notify_counter_channel.receive
+      sleep
     end
 
-    private def scan_files
+    private def watcher(key, files, commands)
+      if key != "server"
+        @notify_channel.receive
+      end
+      if files.empty?
+        commands.each do |command|
+          run_command(command, key)
+        end
+      else
+        loop do
+          scan_files(key, files, commands)
+          @watch_running = true if key == "server"
+          sleep 1
+        end
+      end
+    end
+
+    private def scan_files(key, files, commands)
       file_counter = 0
-      Dir.glob(files) do |file|
-        timestamp = get_timestamp(file)
-        if FILE_TIMESTAMPS[file]? != timestamp
-          if @app_running
-            log "File changed: #{file.colorize(:light_gray)}"
-          end
-          FILE_TIMESTAMPS[file] = timestamp
-          file_counter += 1
+      @file_watcher.scan_files(files) do |file|
+        if @watch_running
+          debug "File changed: #{file}"
         end
+        file_counter += 1
       end
       if file_counter > 0
-        log "Watching #{file_counter} files (server reload)..."
-        start_app
+        debug "Watching #{file_counter} #{key} files"
+        kill_processes(key)
+        commands.each do |command|
+          if key == "server" && command == commands.first?
+            run_build_command(command, commands)
+          else
+            run_command(command, key)
+          end
+        end
       end
     end
 
-    private def stop_all_processes
-      log "Terminating app #{project_name}..."
-      @processes.each do |process|
-        process.kill unless process.terminated?
+    private def check_directories
+      Dir.mkdir_p("bin")
```

Comment:
why does this create `bin/`?

Reply:
Because bin is needed to store the generated binary

Useful when your tasks is like `crystal build src/myapp.cr -o bin/myapp`, so you don't need to execute `mkdir -p` first. `amber watch` already does this, just like `shards build`